hbase-client/src/main/java/org/apache/hadoop/hbase/client/MasterRegistry.java

```diff
@@ -115,20 +129,50 @@
   MasterRegistry(Configuration conf) throws IOException {
     this.hedgedReadFanOut = Math.max(1, conf.getInt(MASTER_REGISTRY_HEDGED_REQS_FANOUT_KEY,
       MASTER_REGISTRY_HEDGED_REQS_FANOUT_DEFAULT));
-    int rpcTimeoutMs = (int) Math.min(Integer.MAX_VALUE,
+    rpcTimeoutMs = (int) Math.min(Integer.MAX_VALUE,
       conf.getLong(HConstants.HBASE_RPC_TIMEOUT_KEY, HConstants.DEFAULT_HBASE_RPC_TIMEOUT));
     // XXX: we pass cluster id as null here since we do not have a cluster id yet, we have to fetch
     // this through the master registry...
     // This is a problem as we will use the cluster id to determine the authentication method
     rpcClient = RpcClientFactory.createClient(conf, null);
     rpcControllerFactory = RpcControllerFactory.instantiate(conf);
-    Set<ServerName> masterAddrs = parseMasterAddrs(conf);
+    // Generate the seed list of master stubs. Subsequent RPCs try to keep a live list of masters
+    // by fetching the end points from this list.
+    populateMasterStubs(parseMasterAddrs(conf));
+    Runnable masterEndPointRefresher = () -> {
+      while (!Thread.interrupted()) {
+        try {
+          // Spurious wake ups are okay, worst case we make an extra RPC call to refresh. We won't
+          // have duplicate refreshes because once the thread is past the wait(), notify()s are
+          // ignored until the thread is back to the waiting state.
+          synchronized (refreshMasters) {
+            refreshMasters.wait(WAIT_TIME_OUT_MS);
+          }
+          LOG.debug("Attempting to refresh master address end points.");
+          Set<ServerName> newMasters = new HashSet<>(getMasters().get());
+          populateMasterStubs(newMasters);
+          LOG.debug("Finished refreshing master end points. {}", newMasters);
+        } catch (InterruptedException e) {
+          LOG.debug("Interrupted during wait, aborting refresh-masters-thread.", e);
```

Comment:
Should this interrupt be accompanied by a shutdown of the managing thread pool? Maybe in an attached exception handler. Or it's okay to leave the pool as abandoned, on the assumption that there was only the single thread in the pool and the whole process is being terminated.

Reply:
Oh, reading below, there is no thread pool. I +1 Viraj's suggestion of a single thread executor service.

Reply:
Let me try again, and ask about the application lifecycle of this thread. It looks like a thread we always want running, unless the Connection is being closed. Now that there's a pool, I'm asking about the lifecycle of the pool, when it's `shutdown` method is called. I'm also asking about the presence of the thread -- is there a reason the thread could be interrupted (and thus terminated) while the connection remains active? There's only one call to submit, and there's nothing that checks for the presence of the thread. So can a connection enter a state where it's still alive with no masterRefreshThread running?

Reply:
>  is there a reason the thread could be interrupted (and thus terminated) while the connection remains active

I don't think so. The only interruptor for this thread should be from the shutdown from the registry#close(), which is called during connection clean up.

> So can a connection enter a state where it's still alive with no masterRefreshThread running?

AFAICT that is not possible. The only way it can break the loop is via interruption.

Reply:
👍